includes/payment-methods/class-wc-everypay-gateway.php

```diff
@@ -42,200 +42,214 @@ class WC_Everypay_Gateway extends WC_Payment_Gateway
-
-	    } catch (Exception $e) {
-		    $error = pll__('An error occurred. Please try again.');
-		    wc_add_notice(esc_html($error), $notice_type = 'error');
-		    (new WC_Everypay_Repository())->save_logs('error', $e->getMessage());
-		    return;
-	    }
-
-    }
-
-    private function complete_order($wc_order)
-    {
-	    $dt = new DateTime("Now");
-	    $timestamp = $dt->format('Y-m-d H:i:s e');
-
-	    $wc_order->add_order_note(pll__('Everypay payment completed at-' .$timestamp));
-	    $wc_order->payment_complete();
-	    $wc_order->get_order();
-	    WC()->cart->empty_cart();
-
-	   return array(
-		    'result' => 'success',
-		    'redirect' => $this->get_return_url($wc_order)
-	    );
-
-    }
+	public function process_payment($order_id)
+	{
+		try {
+			if (isset($_POST['delete_card']) && is_user_logged_in() && $this->tokenization_status == 'yes') {
+				$user_id = get_current_user_id();
+				(new WC_Everypay_Tokenization())->delete_card($_POST['delete_card'], $user_id);
+				return array(
+					'result'   => 'success',
+					'messages' => '<div class=""></div>'
+				);
+			}
+			if (!isset($_POST['everypayToken']) || empty($_POST['everypayToken'])) {
+				$this->renderer->render_iframe(WC()->cart->total, $this->max_installments);
+				exit;
+			}
+			$token = sanitize_text_field($_POST['everypayToken']);
+			unset($_POST['everypayToken']);
+			$wc_order = new WC_Order($order_id);
+			WC_Everypay_Api::setApiKey($this->everypaySecretKey);
+			$payload = $this->create_payload($wc_order, $token);
+
+			if (is_user_logged_in() && (isset($_POST['everypay_save_card']) || isset($_POST['tokenized-card'])) && $this->tokenization_status == 'yes') {
+				(new WC_Everypay_Repository())->save_logs('tokenization_payment', implode(" ", $payload));
+				$user_id = $wc_order->get_user_id();
+				$everypay_tokenization = new WC_Everypay_Tokenization();
+//                @note
+				$response = $everypay_tokenization->process_tokenized_payment($user_id, $payload);
+			} else {
+//                @note
+				(new WC_Everypay_Repository())->save_logs('payment', implode(" ", $payload));
+				$response = WC_Everypay_Api::addPayment($payload);
+			}
+			return $this->complete_order($wc_order, $response['body']['token']);
+		} catch (Exception $e) {
+
+			$error = 'An error occurred. Please try again.';
+			wc_add_notice(esc_html($error), 'error');
+			$response_data = array(
+				'result' => 'failure',
+				'reload' => true,
+				'refresh' => true
+			);
+			echo json_encode($response_data);
+			if ($order_id) {
+				wp_delete_post($order_id, true);
+			}
+			(new WC_Everypay_Repository())->save_logs('error', $e->getMessage());
+			exit;
+		}
+	}
+
+	private function complete_order($wc_order, $token)
+	{
+		$dt = new DateTime("Now");
+		$timestamp = $dt->format('Y-m-d H:i:s e');
+
+		$wc_order->add_order_note('Everypay payment completed at-' . $timestamp);
+//        @note
+        $wc_order->update_meta_data('token', $token);
```

Comment:
Refund fix. On complete order we store the pmt_token, in order to use it on refunds